src/data/finset/noncomm_prod.lean

```diff
@@ -244,4 +280,39 @@ begin
         list.nodup_append_of_nodup sl' tl' h]
 end
 
+section finite_pi
+
+@[to_additive]
+lemma noncomm_prod_single {ι : Type*} [fintype ι] [decidable_eq ι]
+  {M : ι → Type*} [∀ i, monoid (M i)] (x : Π i, M i) :
+  finset.univ.noncomm_prod (λ i, monoid_hom.single M i (x i))
+  (λ i _ j _, by { by_cases h : i = j, { rw h }, { apply pi.mul_single_commute i j h } })
+  = x :=
+begin
+  ext i,
+  apply (finset.univ.noncomm_prod_map (λ i, monoid_hom.single M i (x i)) _
+    (pi.eval_monoid_hom _ i)).trans,
+  rw (finset.insert_erase (finset.mem_univ i)).symm,
+  rw finset.noncomm_prod_insert_of_not_mem',
+  { rw finset.noncomm_prod_eq_one_of_forall_eq_one,
+    { simp, },
+    { intros i h, simp at h, simp [h], } },
+  { simp, },
+end
+
+@[to_additive]
+lemma _root_.monoid_hom.pi_ext {ι : Type*} [fintype ι] [decidable_eq ι]
+  {M : ι → Type*} [∀ i, monoid (M i)] {N : Type*} [monoid N] {f g : (Π i, M i) →* N}
+  (h : ∀ i x, f (monoid_hom.single M i x) = g (monoid_hom.single M i x)) :
+  f = g :=
+begin
+  ext x,
+  rw ← finset.noncomm_prod_single x,
+  rw finset.univ.noncomm_prod_map _ _ f,
+  rw finset.univ.noncomm_prod_map _ _ g,
```

Comment:
You're in the `finset` namespace, so you don't need the prefix here (and elsewhere). Usually it's best to combine multiple `rw`s into a single `rw` too.